src/Todo/DB.hs

```diff
@@ -0,0 +1,54 @@
+module Todo.DB
+    ( TodoDB (..)
+    , DBError (..)
+    , addTodo
+    , addTag
+    , deleteTag
+    , editTag
+    ) where
+
+import Data.HashSet (HashSet)
+import Data.Foldable (toList)
+import qualified Data.HashSet as HSet
+
+import Todo.Model (Todo (..), Tag (..))
+import Data.NESet (NonEmptySet (..))
+import qualified Data.NESet as NESet
+
+data TodoDB = TodoDB
+    { dbTodos :: HashSet Todo
+    , dbTags  :: HashSet Tag
+    } deriving (Eq, Show)
+
+data DBError = SingleTaggedTodo | NotRegisteredTags deriving (Eq, Show)
+
+addTodo :: Todo -> TodoDB -> Either DBError TodoDB
+addTodo todo@Todo {..} TodoDB {..} 
+    | f == mempty = Right TodoDB { dbTodos = HSet.insert todo dbTodos, .. }
+    | otherwise   = Left NotRegisteredTags
+  where
+    f = NESet.filter (\t -> not $ HSet.member t dbTags) todoTags
+
+addTag :: Tag -> TodoDB -> TodoDB
+addTag tag TodoDB {..} = TodoDB { dbTags = HSet.insert tag dbTags, .. }
+
+deleteTag :: Tag -> TodoDB -> Either DBError TodoDB
+deleteTag tag TodoDB {..} = case mapM removeTag $ toList dbTodos of
+    Nothing -> Left SingleTaggedTodo
+    Just x  -> Right TodoDB { dbTodos = HSet.fromList x, ..}
```

Comment:
Probably it's better to name this variable as `todos` to improve code readability.